x-pack/lib/monitoring/internal_pipeline_source.rb

```diff
@@ -13,10 +13,10 @@ class InternalPipelineSource < LogStash::Config::Source::Base
     include LogStash::Util::Loggable
     FEATURE = 'monitoring'
 
-    def initialize(pipeline_config, agent)
+    def initialize(pipeline_config, agent, settings = nil)
       super(pipeline_config.settings)
       @pipeline_config = pipeline_config
-      @settings = LogStash::SETTINGS.clone
+      @settings = settings || LogStash::SETTINGS.clone
```

Comment:
🤔 from what I can tell, this (and the relevant signature change to this method) is only necessary so that specs can work with a modified `LogStash::SETTINGS`, and is not used outside of specs.

Could we use the settings object from `PipelineConfig#settings` (that we already pass to `super`) instead of reaching out to a global constant? Our Specs seem to initialize this with the same `system_settings` that this changeset ends up passing to the new third positional argument of this method.

If we do need to reach out to a global constant at runtime, could this be better handled by mocking the `LogStash::SETTINGS` constant in spec?

Reply:
actually, to simplify, I'd rather remove the `SETTINGS.clone` from `InternalPipelineSource` altogether. I think it will be clearer if this method always require a settings parameter instead and the caller is responsible to provide is with the required settings.

Reply:
Since the `pipeline_config` we are passed already _has_ a `Settings` object, what is the value in shimming a _second_ settings object in? I think requiring two setting objects, accessed in different ways, is an opportunity for bugs (e.g, where our specs mock one but not the other, testing behaviour that _cannot_ happen in real life).

Reply:
the second object was already in with  `@settings = LogStash::SETTINGS.clone` as was previously. My change now at least makes it explicit that a settings object is required (for valid reasons or not) by using a constructor param to set the ivar `@settings = settings` instead of cloning from a global constant inside this `InternalPipelineSource` object which I think is quite obscure. I much prefer to have the caller decide which settings it needs to pass to that constructor, in this case I decided to keep the exact same semantic by doing 
```
LogStash::SETTINGS.set("node.uuid", runner.agent.id)
internal_pipeline_source = LogStash::Monitoring::InternalPipelineSource.new(setup_metrics_pipeline, runner.agent, LogStash::SETTINGS.clone)
```

By doing this it also allows to simply pass config objects for easier testing.

I am not sure that refactoring beyond that by looking into using the  `Settings` object inside the `pipeline_config` should be dealt with in the scope or this PR. My changes here IMO improves a bit on having more strait forward & testable code. I am not sure why a cloned settings is required or not in `InternalPipelineSource` TBH and do not think this needs to be dealt with here.

Reply:
You're right. Your solution doesn't make it worse, only more explicit. I'll resolve this holdover.